### ISSUE TITLE
Absolute file path with ö fails to play audio file - Windows only

### ISSUE THREAD

**Issue by JoshuaD84** [caprica/vlcj#645]
I took a flac file and copied it to four locations:

[workingDirectory]/o.flac
[workingDirectory]/ö.flac
D:\test\o.flac
D:\test\ö.flac

I then pointed VLCJ at each file.  It **fails** to play when targeting the absolute path that has the ö in it.   It works in the other three cases.

This bug does not appear to occur in GNU/Linux based system.  It is occurring on Windows 7. I have not tested Windows 8 & 10. 

I do not get any errors in any of the test cases; I get the same output whether I hear music or not. 

> 69 [main] INFO uk.co.caprica.vlcj.Info - vlcj: 3.10.1
> 69 [main] INFO uk.co.caprica.vlcj.Info - java: 1.8.0_181 Oracle Corporation
> 70 [main] INFO uk.co.caprica.vlcj.Info - java home: C:\Program Files\Java\jdk1.8.0_181\jre
> 70 [main] INFO uk.co.caprica.vlcj.Info - os: Windows 7 6.1 amd64
> 104 [main] INFO uk.co.caprica.vlcj.binding.LibVlcFactory - vlc: 3.0.3 Vetinari, changeset 3.0.3-1-0-gc2bb759264
> 104 [main] INFO uk.co.caprica.vlcj.binding.LibVlcFactory - libvlc: D:\programming\workspace-win\Hypnos-Win\stage\lib\win\vlc\libvlc.dll

Is there a workaround I can use for this? 

Edit: further testing shows this occurs for mp3s as well, so it does not appear to be a flac-specific issue. 

Here a MCVE: 

```
import java.nio.file.Paths;
import com.sun.jna.NativeLibrary;
import uk.co.caprica.vlcj.component.AudioMediaPlayerComponent;
import uk.co.caprica.vlcj.player.MediaPlayer;
import uk.co.caprica.vlcj.runtime.RuntimeUtil;

public class VLCJFlacError {
	
	final static String[] targetFile = {
			"D:\\test\\o.flac",   //works
			"D:\\test\\ö.flac",   //does not work
			"o.flac",             //works
			"ö.flac"              //works
			"D:\\test\\o.mp3",   //works
			"D:\\test\\ö.mp3",   //does not work
	};
	
	final static int currentTestIndex = 3;
	
	final static String vlcLibPath = "stage/lib/win/vlc";
	
	public static void main( String[] args ) throws Exception {
		
		String nativeVLCLibPath = Paths.get( vlcLibPath ).toAbsolutePath().toString();
		
		NativeLibrary.addSearchPath( RuntimeUtil.getLibVlcLibraryName(), nativeVLCLibPath );
		AudioMediaPlayerComponent vlcComponent = new AudioMediaPlayerComponent();
		MediaPlayer mediaPlayer = vlcComponent.getMediaPlayer();
		
		mediaPlayer.playMedia( targetFile[ currentTestIndex ]  );
		
		while ( true ) { //lazy stay alive so the media can play
			Thread.sleep ( 50 );
		}
	}
}
```

**Comment by JoshuaD84** [comment]
Perhaps these javadoc notes may be part of the issue: http://caprica.github.io/vlcj/javadoc/3.10.1/index.html

> 
> Encode, if needed, a local file MRL that may contain Unicode characters as a file URL with "percent" encoding.
> This method deals only with the special case of an MRL for a local file name containing Unicode characters. Such MRLs must be encoded as file URLs, by adding a "file://" prefix before percent-encoding the filename.
> 
> Without this, vlc will not be able to play the file since it is using native API that can not handle unencoded Unicode characters.
> 
> This method does not deal with any MRLs that are URLs since Unicode characters are forbidden by specification for any URL.
> 
> What this means in practical terms is that if an MRL is specified that contains a "scheme" like "http", or "file" then that MRL will not be encoded by this method, even if it contains Unicode characters. This situation, if it arises, is considered a client application vaildation failure.

**Comment by JoshuaD84** [comment]
Sweet, I found a workaround, using some code from here: https://github.com/caprica/vlcj/blob/master/src/main/java/uk/co/caprica/vlcj/player/MediaResourceLocator.java

For some reason, that doesn't seem to be catching the character 'ö'. 

I took the key code from there and made this, which works for my program and use case:

```
	private String encodeFileToVLCString ( String filename ) {
		//Assumes filename is an absolute file location. 
        return new File(filename).toURI().toASCIIString().replaceFirst("file:/", "file:///");
	}
```

**Comment by caprica** [comment]
For me, with the character you posted ö, containsUnicode() returns true. So the bug is not there.

**Comment by caprica** [comment]
So in your failing test cases, vlcj skips any escaping of the Unicode characters because there is a piece of code that checks to see if the supplied MRL is a URL or a local file, and if a URL can not be created from the MRL the result is that no encoding is required.

Escaping of Unicode characters is required for URLs because they are forbidden by specification.

This is basically what the Javadoc is saying.

So for your test cases this leaves us with vlcj passing the unescaped string to LibVLC.

On my Linux setup, this works just fine. Filenames with Unicode characters are handled without any difficulty - it just works.

As is completely unsurprising, frankly, this does not work with Windows filenames. It therefore appears that on Windows (I don't think you say which version you use) the un-escaped string can *not* be used, I don't know where the problem is, it may be inside the JVM, somewhere in JNA, or infact it may be inside LibVLC itself.

The reason that "ö.flac" works and "D:\\test\\ö.flac" does not work is that a valid URL can be inferred from the former and so is escaped.

**Comment by caprica** [comment]
As to the solution...

On the face of it your workaround seems reasonable, it is essentially escaping the MRL if it contains any Unicode characters whether or not the MRL is a URL or a local filename. If this is known to work in all cases then probably a change should be made in this code in vlcj - this change would only be effected if the runtime were Windows since as I stated above it works just fine on Linux.

**Comment by caprica** [comment]
#467

**Comment by caprica** [comment]
This is a confounding problem. Having recently tested this on my own Windows installation, all those filename variants just worked - i.e. before making any code changes.

I am not opposed to changing vlcj to try and accommodate this issue, but the suggestion is there is something else wrong as these filenames with unicode characters (relative or absolute, makes no difference) all just work.

**Comment by caprica** [comment]
Hopefully fixed by 7e55fdcb8ffe77aec3ec49075d67f955a16b7e28.

I say "hopefully" because in my own Windows environment I never encountered the problem, it has always worked, so I don't really have the means to test the failure-path properly.

**Comment by caprica** [comment]
Just for the record, I was finally able to reproduce this in my Windows environment. It was bothering me that I couldn't make it fail.

The trick to making it fail, for me at least, was to specify -Dfile.encoding=ASCII when starting the JVM.

When I specify that encoding, without the fix discussed in this issue it will fail.

**Comment by laim0nas100** [comment]
TLDR:
Unicode symbol è  (\u00E8) seems to be an issue.

I've checked several other symbols, and only that one seems to be problematic, but probably there could be more.

Not sure if this is worth a separate thread, because it seems to be related.
Also Windows 10.
Java version seems to not be relevant (tested 1.8, 11, 14).
VLCJ version is 4.7.0
VLC version 3.0.11

I've checked the encoding function and it is correct, so the error happens somewhere deeper.

I've also tested simple command line invocation from java using ProcessBuilder with vlc.exe, and it usually works, except when the è symbol is written as \u00E8.
Then I get an error from libVlc itself:
[000002298de80200] filesystem stream error: cannot open file D:\Dev\Java\Workspace\VLCJbugreport\Ã¨.mp4 (No such file or directory)
Which is the wrong path, so the encoding messes up somewhere.
This leads me to believe the error could be not in vlcj but in libVlc itself, so just be aware that the encoding problem is not fully fixed in some niche cases.

Here's the code I used for testing (-imports):

    

    public class Main {

        public static final boolean playAll = true;
        public static final boolean useCMD = false;

        public static void main(String[] args) throws InterruptedException, IOException {
            final EmbeddedMediaPlayerComponent comp = new EmbeddedMediaPlayerComponent();
            JFrame f = new JFrame();
            f.setContentPane(comp);
            f.setBounds(200, 200, 800, 600);
            f.addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    comp.release();
                    System.exit(0);
                }
            });

            f.setVisible(true);
            String dir = "D:\\test\\vlcj\\";
            //every file here is readable
            String[] ref = {
                "Paris Blohm & Wasback - They Say (feat. Krigarè).mp4",// fail
                "(feat. Krigarè).mp4",// fail
                "(è.mp4",// fail
                "(.è).mp4",// fail
                ".è.mp4",// fail

                "\u00E8.mp4",//fail on CMD also
                "èèè.mp4",//fail
                "(feat.à).mp4", //ok
                "à).mp4",//ok
                "ì).mp4", //ok
                "♥♥♥.mp4",//ok
                "_È.mp4", //ok
                "(_È).mp4", //ok
                "(._È).mp4" //ok
            };

            EmbeddedMediaPlayer mediaPlayer = comp.mediaPlayer();
            mediaPlayer.events().addMediaEventListener(new MediaEventAdapter() {

                @Override
                public void mediaStateChanged(Media media, State newState) {
                    System.out.println("mediaStateChanged:" + media.info().mrl() + " " + newState);
                }

            });
            mediaPlayer.events().addMediaPlayerEventListener(new MediaPlayerEventAdapter() {

                @Override
                public void error(MediaPlayer mediaPlayer) {
                    System.out.println("Error");
                }

                @Override
                public void mediaPlayerReady(MediaPlayer mediaPlayer) {
                    System.out.println("Ready");
                }
            });

            if (playAll) {
                for (String file : ref) {
                    testPlay(mediaPlayer, dir + file);
                }
            } else {
                testPlay(mediaPlayer, "è.mp4");//local path fails, but ok using CMD
                testPlay(mediaPlayer, "\u00E8.mp4");//this local path fails, even with CMD
            }
            System.exit(0);

        }

        public static void testPlay(MediaPlayer player, String toPlay) throws InterruptedException, IOException {
            boolean canRead = new File(toPlay).canRead();

            System.out.println("Readable:" + canRead + " " + toPlay);
            if (useCMD) {
                testPlayCMD(toPlay);
                return;
            }
            boolean ok = player.media().prepare(toPlay); // this is always true for some reason
            Thread.sleep(500);
            player.controls().play();
            Thread.sleep(1000);
            System.out.println("Playable:" + player.status().isPlayable() + " Playing:" + player.status().isPlaying() + " " + toPlay);
            Thread.sleep(2000);
            player.controls().stop();
        }

        public static void handleStream(Process process, Function<Process, InputStream> streamGet, Consumer<String> lineCons) {
            Runnable handle = () -> {
                BufferedReader reader = new BufferedReader(new InputStreamReader(streamGet.apply(process)));
                try {
                    while (process.isAlive()) {
                        String line = reader.readLine();
                        while (line != null) {
                            lineCons.accept(line);
                            line = reader.readLine();
                        }
                    }
                } catch (IOException ex) {
                }
            };
            new Thread(handle).start();
        }

        public static void testPlayCMD(String toPlay) throws IOException, InterruptedException { 
            System.out.println("Play with cmd");
            ProcessBuilder processBuilder = new ProcessBuilder()
                    .command("vlc.exe", toPlay);
            Process process = processBuilder.start();
            handleStream(process, Process::getInputStream, System.out::println);
            handleStream(process, Process::getErrorStream, System.err::println);
            Thread.sleep(3000);
            process.destroy();
        }
    }

**Comment by caprica** [comment]
Very interesting, but this is so strange. I doubt very much there is anything particular special about the failing characters, it can only be a wrong encoding _somewhere_- what I mean is what possible code could be fixed in vlcjLibVLC for such a bug?

**Comment by laim0nas100** [comment]
From my tests, I think the issue lies in the way Windows handles unicode in command line.
I could not find a way to pass a file path with a problematic symbol through arguments (i.e. java -jar myJar.jar "some crazy file path with symbols èèè"), that would later be readable (new File(path).canRead()). 
If the issue lies somewhere, a way to get argument passing working should shed light on this problem and could be related.
Changing codepages did not help.

If there is something "fixable" in vlcj, it should report the error when "prepare" method is called as false, if libVlc fails to find a file in question. Obviously, the error still remains, but my suspicion is that there is some weird non-standard exception in WIndows unicode handling, that makes this error out of vlcj or libVlc "jurisdiction".
